kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/blockchain/BlockchainSourceImpl.java

```diff
@@ -156,48 +187,78 @@ public void onError(final Exception e) {
                     @Override
                     public void run() {
                         if (e instanceof AccountNotFoundException) {
-                            callback.onResponse(0);
+                            callback.onResponse(balance.getValue());
                         } else {
                             callback.onFailure(e);
                         }
                     }
                 });
-                Log.d(TAG, "getCurrentBalance onError: " + e.getMessage());
+                Log.d(TAG, "getBalance onError: " + e.getMessage());
             }
         });
     }
 
-    private void setBalance(int balanceValue) {
-        balance.setValue(balanceValue);
-        local.setBalance(balanceValue);
+    @VisibleForTesting
+    void setBalance(final Balance balanceObj) {
+        BalanceUpdate balanceUpdate = balance.getValue();
+        // if the values are not equals so we need to update,
+        // no need to update for equal values.
+        if (balanceUpdate.getAmount().compareTo(balanceObj.value()) != 0) {
+            Log.d(TAG, "setBalance: Balance changed, should get update");
+            balanceUpdate.setAmount(balanceObj.value());
+            balance.postValue(balanceUpdate);
+            local.setBalance(balanceObj.value().intValue());
+        }
     }
 
     @Override
-    public void addBalanceObserver(@NonNull Observer<Integer> observer) {
+    public void addBalanceObserver(@NonNull Observer<BalanceUpdate> observer) {
         balance.addObserver(observer);
         observer.onChanged(balance.getValue());
     }
 
     @Override
-    public void removeBalanceObserver(@NonNull Observer<Integer> observer) {
+    public void addBalanceObserverAndStartListen(@NonNull Observer<BalanceUpdate> observer) {
+        addBalanceObserver(observer);
+        Log.d(TAG, "addBalanceObserverAndStartListen: " + balanceObserversCount.get());
+        if (balanceObserversCount.getAndIncrement() == 0) {
```

Comment:
is it for thread safety? (are you expecting/allowing to call this from multiple threads)
if so there might be a race condition here, the `startBalanceListener` is a critical section that should be locked. see also below at `decrementPaymentCount`

Reply:
why? balanceObserversCount is static volatile